plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Simplify.hs

```diff
@@ -74,3 +89,10 @@ simplifyTerm opts = simplifyNTimes (_soMaxSimplifierIterations opts)
     caseOfCase' = case eqT @fun @DefaultFun of
         Just Refl -> caseOfCase
         Nothing   -> id
+
+    cseStep :: Int -> Term name uni fun a -> m (Term name uni fun a)
+    cseStep _ = case (eqT @name @Name, eqT @uni @PLC.DefaultUni) of
```

Comment:
Hardcoding `Name` here is kinda weird. Any name should work as long as it has a `TermUnique`.

Reply:
I think the current implementation of CSE only works correctly for `Name`. We can't just do the substitutions that we do for `DeBruijn` terms.

Reply:
What about requiring `HasUnique`? That's usually what we do in some situations, and it's properly polymorphic instead of this abomination.

Reply:
> What about requiring HasUnique?

I also need `freshName`...

Reply:
Okay, then just monomorphise to `Name` and `TyName`. The PIR compiler in fact runs on those types, and if that's the truth you might as well say it.

Reply:
Bummer, and I was looking forward to generalize PIR to work with any name/tyname. But of course this i s noot the only place where we do fix to Name/TyName. Other places that I can think of is the PIR typechecker and the rewrite rules